### ISSUE TITLE
Optionally support a FIPS-approved `CryptoProvider`

### ISSUE THREAD

**Issue by ctz** [rustls/rustls#1540]
This is something we plan to do in the future, and will imply some minor changes to the `CryptoProvider` interface.

This new provider should:

- only include key exchange mechanisms that are backed by a FIPS-approved implementation
- only include cipher suites that are entirely made up of FIPS-approved parts
- only support authentication signing algorithms that are backed by a FIPS-approved implementation
- only support certificate verification algorithms that are backed by a FIPS-approved implementation

**Comment by djc** [comment]
Can you describe (high-level) what changes you expect to be necessary to the `CryptoProvider` interface itself?

**Comment by ctz** [comment]
Because NIST validate the [TLS 1.2](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-135r1.pdf) and 1.3 key derivation functions, it is necessary to use the validated implementations rather than our own instantiated with HMAC (even though that HMAC implementation itself would also be validated).

The rough change I'm envisioning is something like:

```
  pub struct Tls12CipherSuite {
- pub hmac_provider: &'static dyn crypto::hmac::Hmac,
+ pub prf_provider: &'static dyn crypto::tls12::Prf,
}
```

(Same for TLS1.3. There are no other uses of `hmac_provider` in either case.)

Then we'd export a struct that impls `crypto::tls12::Prf`, such that the code in `provider-example` would change like:

```
-         hmac_provider: &hmac::Sha256Hmac,
+         prf_provider: &rustls::crypto::tls12::HmacPrf(&hmac::Sha256Hmac)
```

This means non-FIPS providers don't need to implement the TLS1.2 PRF/HKDF directly. That's especially important because (it seems) nobody can agree what the HKDF interface should look like.

**Comment by djc** [comment]
Should we just get that change over with?

**Comment by briansmith** [comment]
> prf_provider: &rustls::crypto::tls12::HmacPrf(&hmac::Sha256Hmac)

Not sure if this is the place to discuss this, but it would be preferable to allow the crypto provider to do the key agreement and calculate the keyblock in a single atomic operation, if practical, so that the raw result of the key agreement (especially RSA decryption, but also ECDH or X25519) can be held only within the crypto provider and not exposed to Rustls.

**Comment by janrueth** [comment]
I prototyped a boringssl-based crypto provider: https://github.com/janrueth/boring-rustls-provider
Might be of interest with regard to FIPS.

**Comment by ctz** [comment]
Very interesting -- thanks!

**Comment by cetanu** [comment]
Sorry to bump this, was this closed because the plan is to migrate to `rustls`, and rely upon that dependency allowing the use of the `boring` ssl crate, thereby solving FIPS compliance?

**Comment by djc** [comment]
> Sorry to bump this, was this closed because the plan is to migrate to `rustls`, and rely upon that dependency allowing the use of the `boring` ssl crate, thereby solving FIPS compliance?

Given that this is an issue in the rustls issue, not sure how to make sense of this question?

I think this was closed because we merged #1780, which changes rustls' default crypto providers to aws-lc-rs, which can provide FIPS assurances if requested.
